### PR TITLE
Add OpInfo based meta tensor tests

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76905

https://github.com/pytorch/pytorch/pull/75994 was taking too long to
ship so I extracted out the CrossRef gadget and had it run on a simple
OpInfo invocation only.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: test/test_meta.py [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+                    meta_rs = func(*meta_args, **meta_kwargs)
+            except Exception as e:
+                suppress = False
+                if isinstance(e, NotImplementedError):
+                    m = RE_NOT_IMPLEMENTED_MSG.search(e.args[0])
+                    if m and m.group(1) not in ("aten::_efficientzerotensor", "aten::view_as_real"):
+                        suppress = True
+                if not suppress:
+                    raise RuntimeError(f"""\
+failed to run: {func}(
```
Comment (anjali411) [inline test/test_meta.py:786]: indentation

**Review by anjali411 — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -9603,6 +9603,7 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                # Reference: https://github.com/pytorch/pytorch/issues/50747
                DecorateInfo(unittest.skip("Skipped!"), 'TestCommon', 'test_variant_consistency_eager',
                             dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.float16)),
+               DecorateInfo(unittest.expectedFailure, 'TestMeta', 'test_meta', dtypes=(torch.bool,)),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:9606]: Nice xfails

**Comment by ezyang** [comment]
@mruberry So a question: for stuff like this, I could have instead found a test where we're running forwards, and smooshed my test into that one. The tests are less isolated but they run faster (because instead of running the stock CPU/CUDA implementation twice, I instead only run it once).

File: test/test_meta.py [hunk 3]
```diff
@@ -0,0 +1,846 @@
...
+
+import functools
+import re
+from functools import partial
+import unittest
+import warnings
+
+RE_NOT_IMPLEMENTED_MSG = re.compile(r"Could not run '([^']+)' with arguments ")
+
+meta_exclude_set = {
```
Comment (mruberry) [inline test/test_meta.py:28]: holy exclude set, Batman!

File: test/test_meta.py [hunk 4]
```diff
@@ -0,0 +1,846 @@
...
+    torch.tensor,
+    torch.zeros_like,
+    torch.randn_like,
+    torch.sparse_csr_tensor,
+    torch._sparse_coo_tensor_unsafe,
+    torch._sparse_csr_tensor_unsafe,
+    torch._validate_sparse_csr_tensor_args,
+}
+
+class MetaCrossRefMode(torch.overrides.TorchFunctionMode):
```
Comment (mruberry) [inline test/test_meta.py:657]: Comment pls

File: test/test_meta.py [hunk 5]
```diff
@@ -0,0 +1,846 @@
...
+
+        return rs
+
+class TestMeta(TestCase):
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyNativeDeviceTypes
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(op_db)
+    def test_meta(self, device, dtype, op):
```
Comment (mruberry) [inline test/test_meta.py:832]: Comment describing this test, too, would be nice

**Review by mruberry — APPROVED** [review]
Cool! A few inline suggestions but looks pretty good from a test framework perspective

File: test/test_meta.py [hunk 6]
```diff
@@ -0,0 +1,846 @@
...
+    torch.Tensor.addcmul,  # MISSING aten::_local_scalar_dense
+    torch.Tensor.angle,  # MISSING aten::angle
+    torch.Tensor.argsort,  # MISSING aten::sort
+    torch.Tensor.bincount,  # MISSING aten::bincount
+    torch.Tensor.cholesky,  # MISSING aten::cholesky
+    torch.Tensor.cholesky_inverse,  # MISSING aten::cholesky_inverse
+    torch.Tensor.cholesky_solve,  # MISSING aten::_cholesky_solve_helper
+    torch.Tensor.clamp,  # MISSING aten::clamp.Tensor
+    torch.Tensor.clamp_,  # MISSING aten::clamp.Tensor_out
+    torch.Tensor.clamp_max,  # MISSING aten::clamp_max.out
```
Comment (ngimel) [inline test/test_meta.py:49]: Should be fixed by #76874 for scalar overload at least

**Comment by mruberry** [comment]
> @mruberry So a question: for stuff like this, I could have instead found a test where we're running forwards, and smooshed my test into that one. The tests are less isolated but they run faster (because instead of running the stock CPU/CUDA implementation twice, I instead only run it once).

Totally! This comes up a lot (our ops get called on the same sample inputs SO MANY TIMES!). And combining tests would make the test suite faster, but it would also impact test granularity. Like if the meta portion of a test failed then what about the dtypes part, or the autograd part, etc. What would be really interesting is if we could use subtest-like functionality to have these larger tests that preserved their computed artifacts for reuse across a variety of subtests, but we'd need to ensure the subtests could succeed or fail independently, and that we could still apply our xfail/skip/tolerance/etc. decorators to the subtests, too.

File: test/test_meta.py (hunk 1 above)
Comment (ezyang) [reply test/test_meta.py:793]: actually unindenting triple blocks like this is idiomatic, because indenting them will result in indentation in the underlying string

File: test/test_meta.py [hunk 7]
```diff
@@ -0,0 +1,846 @@
...
+
+class TestMeta(TestCase):
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyNativeDeviceTypes
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(op_db)
+    def test_meta(self, device, dtype, op):
+        func = op.get_op()
+        if func in meta_exclude_set:
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:834]: I used to do this, but this makes the test suite run *a lot* longer, like 4x longer (because throwing exceptions is expensive). As a compromise, I can add an optional mode to do it this way, but not the default.

[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_meta.py:834]: Yeah I understand it takes longer. If you don't want to do it let's not bother, we can add a mode later if we really want to

File: test/test_meta.py (hunk 6 above)
Comment (ngimel) [reply test/test_meta.py:49]: And now tensor overloads are getting fixed too.

File: test/test_meta.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py]: So, when I'm doing OpInfo tests only, it's tolerable. So I've turned on expected failures, and shuffled things around so that I can NOT do that for crossref testing.

File: aten/src/ATen/native/TensorShape.cpp [hunk 8]
```diff
@@ -201,6 +201,11 @@ Tensor& set_storage_cpu_(Tensor& result, Storage storage, int64_t storage_offset
   return result;
 }
 
+Tensor& set_(Tensor& result, const Tensor& storage, int64_t storage_offset, IntArrayRef size, IntArrayRef stride) {
```
Comment (albanD) [inline aten/src/ATen/native/TensorShape.cpp:204]: nit: storage -> source

File: test/test_meta.py [hunk 9]
```diff
@@ -0,0 +1,900 @@
...
+    torch.Tensor.isfinite,  # MISSING aten::abs.out
+    torch.isfinite,  # MISSING aten::abs.out
+    torch.diff,  # MISSING aten::logical_xor.out
+}
+
+# These are fine in OpInfo tests, but triggered errors in full test suite
+# crossref testing, which means there is probably not enough coverage from
+# OpInfo.  Patch in https://github.com/pytorch/pytorch/pull/75994 and find
+# out where these fails come from.
+suspicious_exclude_set = {
```
Comment (albanD) [inline test/test_meta.py:479]: nit: not used? do we want to keep it?

**Review by albanD — APPROVED** [review]
SGTM

File: test/test_meta.py (hunk 9 above)
Comment (ezyang) [reply test/test_meta.py:479]: Yes I do, I'll use it for real crossref testing later
